src/core/qgsabstractdatabaseproviderconnection.cpp

```diff
@@ -226,6 +227,101 @@ QgsAbstractDatabaseProviderConnection::TableProperty QgsAbstractDatabaseProvider
                                         .arg( name, schema ) );
 }
 
+QList<QgsVectorDataProvider::NativeType> QgsAbstractDatabaseProviderConnection::nativeTypes() const
```

Comment:
This method is rather ugly -- wouldn't it be much less work/more reliable to move the native types methods from the various providers to a static function and use that from the connection implementations?

Reply:
@nyalldawson I've thought about this a lot, the reason why it needs to be dinamic is the OGR provider: being it a multi-storage provider the nativeTypes function can only be executed on a layer instance because it needs to know the GDAL/OGR driver.

The alternative is to implement it in the metadata/connections API but in that case there would be a duplication of code to return the hardcoded list of native types, and I'd like to avoid that.

The assumption here is that the user is going to create a new table so he must have permissions to do so or it will fail post-mortem when the dialog is accepted anyway so we should be on the safe side (functionally speaking).

Reply:
Hm - good point. But I still think it's a generally ugly approach for something like postgres. The postgres situation is rather simple -- we just extract the lump of code from qgspostgresprovider.cpp lines 228->270 and move that to a static function, which we call from both qgspostgresprovider.cpp and qgspostgresproviderconnection.cpp

Similar applies to mssql and spatialite. Really OGR is the odd one out here, so I'd only do the fallback approach for that provider.

Reply:
If you are ok with duplicating code, I'll go that way. Another point I was also thinking at is the possibility that different endpoints of the same provider support different data types for example if there is a particular postgres extension installed. That would be another reason to keep it dynamic.

Reply:
I've just checked pg provider and there is already code that expose different fields for different versions of postgres.

Reply:
Why is it duplicating code? It's moving the existing code to a re-usable standalone function...

Reply:
1. we cannot make `QList<QgsVectorDataProvider::NativeType> QgsVectorDataProvider::nativeTypes() const` static because it returns different values depending on the driver for OGR and depending on PG version and extensions for postgres
2. we cannot make it static in `QgsAbstractDatabaseProviderConnection` either for the same reason, unless ...

... if we settled for a subset of supported field types we could make it static in `QgsAbstractDatabaseProviderConnection` but that would mean to hardcode the list of native types in the function implementation, this is duplication of code because the "sister" non-static function `QgsVectorDataProvider::nativeTypes()` would need to return a similar (but dynamic) list.

This is not only duplication of code but we also risk that they eventually get out of sync.

Reply:
static methods done!